### ISSUE TITLE
how to disable the ema during training or inference

### ISSUE THREAD

**Issue by Henryplay** [ultralytics/ultralytics#20115]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello partner, this is my first time using YOLO and I would like to know how to disable the EMA function during inference or training by not using EMA directly. I have found that the weight of the onnx transferred is inconsistent with the weight of the PT model, so I would like to rule out the possibility of EMA.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Henryplay, thank you for your interest in Ultralytics 🚀! We recommend checking out our [Docs](https://docs.ultralytics.com/) for detailed information, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples, which might help address your question.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us better understand and debug the issue.

For your specific ❓ question about disabling EMA during training or inference, please provide additional details, such as the commands or scripts you are using, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/). This will help us assist you more effectively.

Feel free to join the Ultralytics community for further discussions:
- Real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- In-depth discussions: [Discourse](https://community.ultralytics.com/)
- Community threads: [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Make sure you are using the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure the issue is not already resolved:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date verified environments (with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests validate correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response, but rest assured, an Ultralytics engineer will assist you soon! 😊

**Comment by Henryplay** [comment]
# I should provide you with more detailed information  
##### I use the following code to verify the consistency of the weights between the PT model and the ONNX model
```
import torch
import onnx
import numpy as np
import onnxruntime as ort
from torch import nn

def compare_models(pytorch_model, onnx_model_path, num_layers=2):

    pytorch_params = {}
    for name, param in pytorch_model.named_parameters():
        pytorch_params[name] = param.detach().cpu().numpy()

    print("\n[PyTorch model parameter]")
    count = 0
    for name, param in pytorch_params.items():
        if count >= num_layers:
            break
        print(f"layer name: {name}")
        print(f"shape: {param.shape}")
        print(f"param(5): {param.flatten()[:5]}")
        print("---")
        count += 1

    # 3. load onnx model
    onnx_model = onnx.load(onnx_model_path)

    print("\n[ONNX model parameters]")
    count = 0
    for initializer in onnx_model.graph.initializer:
        if count >= num_layers:
            break

        name = initializer.name
        shape = [dim for dim in initializer.dims]

        if initializer.data_type == 1:  # FLOAT
            data = np.frombuffer(initializer.raw_data, dtype=np.float32).reshape(shape)
            print(f"layer name: {name}")
            print(f"shape: {shape}")
            print(f"param(5): {data.flatten()[:5]}")
            print("---")
            count += 1

def main():
    from ultralytics import YOLO
    pytorch_model = YOLO("./yolo11n.pt")
    pytorch_model.export(format="onnx", imgsz=640)
    # yolo11n.onnx is the converted onnx by yolo11n.pt
    compare_models(pytorch_model, "yolo11n.onnx", num_layers=1)

if __name__ == "__main__":
    main()

```
##### and the printed information is
```
Ultralytics 8.3.91 🚀 Python-3.8.20 torch-1.9.1+cu111 CPU (Intel Xeon E5-2690 v4 2.60GHz)
YOLO11n summary (fused): 100 layers, 2,616,248 parameters, 0 gradients, 6.5 GFLOPs

PyTorch: starting from 'yolo11n.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (5.4 MB)

ONNX: starting export with onnx 1.17.0 opset 12...
ONNX: slimming with onnxslim 0.1.49...
ONNX: export success ✅ 3.9s, saved as 'yolo11n.onnx' (10.2 MB)

Export complete (4.2s)
Results saved to /data1/zpsun/workspace/qrcode/ultralytics
Predict:         yolo predict task=detect model=yolo11n.onnx imgsz=640  
Validate:        yolo val task=detect model=yolo11n.onnx imgsz=640 data=/usr/src/ultralytics/ultralytics/cfg/datasets/coco.yaml  
Visualize:       https://netron.app

[PyTorch model parameter]
layer name: model.model.0.conv.weight
shape: (16, 3, 3, 3)
param(5): [   -0.13306   -0.018738     0.17236    -0.33691   -0.044312]
---

[ONNX model parameters]
layer name: model.0.conv.weight
shape: [16, 3, 3, 3]
param(5): [    -2.7524    -0.38761      3.5656     -6.9695    -0.91664]
---
```
What a miraculous thing!

**Comment by Henryplay** [comment]
I have disabled the ema in the EMA class,but i don't know if it's useful

---- Replied Message ----
| From | Mohammed ***@***.***> |
| Date | 04/10/2025 20:25 |
| To | ultralytics/ultralytics ***@***.***> |
| Cc | Henry ***@***.***>,
Mention ***@***.***> |
| Subject | Re: [ultralytics/ultralytics] how to disable the ema during training or inference (Issue #20115) |

I have found that the weight of the onnx transferred is inconsistent with the weight of the PT model, so I would like to rule out the possibility of EMA.

How did you check this?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***>

Y-T-G left a comment (ultralytics/ultralytics#20115)

I have found that the weight of the onnx transferred is inconsistent with the weight of the PT model, so I would like to rule out the possibility of EMA.

How did you check this?

**Comment by glenn-jocher** [comment]
Hi @Henryplay,

The weight difference you're seeing is expected and not necessarily an EMA issue. ONNX export often scales weights for optimization during format conversion. The `onnxslim` process you can see in your logs (`ONNX: slimming with onnxslim 0.1.49...`) performs various optimizations that can alter weight representation.

To properly disable EMA during training, you can add `ema=False` to your training arguments:

```python
model.train(data="coco8.yaml", epochs=3, ema=False)
```

For already trained models, if you want to export the base model without EMA influence:
```python
model = YOLO("yolo11n.pt")
# Access the underlying PyTorch model directly instead of the EMA version
model.model = model.model.model if hasattr(model.model, "model") else model.model
model.export(format="onnx")
```

Rather than comparing raw weights, I'd recommend validating functional equivalence by comparing model predictions on the same inputs between PyTorch and ONNX versions.

**Comment by Henryplay** [comment]
hello, @glenn-jocher  
Thank you for your reply. Instead of using the export method of the YOLO class, I wrote a simple code to load the model and read the weights. After exporting onnx, I compared and found that the weight values of the PT model and onnx model were consistent. Although I don't know why, I will post the simple code next Monday and we will analyze possible reasons together.

**Comment by Y-T-G** [comment]
It's probably due to the batch norm weights getting fused during export which is normal.
